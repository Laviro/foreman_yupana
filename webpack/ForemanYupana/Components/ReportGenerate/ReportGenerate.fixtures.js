import { noop } from 'patternfly-react';

export const exitCode = 'exit 0';
export const logs = ['No running process'];
export const completed = 0;
export const restartProcess = noop;
export const error = null;
export const processScheduledTime = 'some-time';
export const props = {
  exitCode,
  logs,
  completed,
  restartProcess,
  error,
};
