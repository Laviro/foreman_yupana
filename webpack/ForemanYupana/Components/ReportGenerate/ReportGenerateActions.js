import {
  REPORT_GENERATE_LOGS_POLLING,
  REPORT_GENERATE_START,
  REPORT_GENERATE_STOP,
  REPORT_GENERATE_FINISH,
} from './ReportGenerateConstants';

export const fetchLogs = () => dispatch => {
  // TODO: Add API call here
  dispatch({
    type: REPORT_GENERATE_LOGS_POLLING,
    payload: {
      logs: window.generatingLogs,
      completed: window.generatingPercentage,
    },
  });
};

export const startProcess = () => dispatch => {
  // TODO: Add API call here
  window.generatingLogs = [
    'Generating...',
    'Hosts: 20',
    'writing host 1/20',
    'writing host 2/20',
  ];

  const logs = window.generatingLogs;
  const processID = setInterval(() => {
    const lastLog = logs[logs.length - 1];
    const splittedWords = lastLog.split(' ');
    const lastWord = splittedWords.pop();
    let amount;
    let total;
    if (lastWord.includes('/')) {
      const splittedAmounts = lastWord.split('/');
      amount = Number(splittedAmounts[0]);
      total = Number(splittedAmounts[1]);
    }
    if (amount === total) {
      clearInterval(processID);
      dispatch({
        type: REPORT_GENERATE_FINISH,
        payload: { status: 'finished' },
      });
      return;
    }
    const nextAmount = amount + 1;
    const newLog = `${splittedWords[0]} ${
      splittedWords[1]
    } ${nextAmount}/${total}`;
    window.generatingLogs.push(newLog);
    window.generatingPercentage = (nextAmount * 100) / total;
  }, 1500);

  dispatch({
    type: REPORT_GENERATE_START,
    payload: {
      processID,
      logs: window.generatingLogs,
      completed: window.generatingPercentage,
      status: 'running',
    },
  });
};

export const stopProcess = processID => dispatch => {
  // TODO: Add API call here
  clearInterval(processID);
  dispatch({
    type: REPORT_GENERATE_STOP,
    payload: { logs: ['No running process'], completed: 0, status: 'stopped' },
  });
};
