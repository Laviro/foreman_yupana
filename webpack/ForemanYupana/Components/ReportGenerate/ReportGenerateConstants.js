export const REPORT_GENERATE_LOGS_POLLING =
  'YUPANA/REPORT_GENERATE_LOGS_POLLING';
export const REPORT_GENERATE_START = 'YUPANA/REPORT_GENERATE_START';
export const REPORT_GENERATE_STOP = 'YUPANA/REPORT_GENERATE_STOP';
export const REPORT_GENERATE_FINISH = 'YUPANA/REPORT_GENERATE_FINISH';
