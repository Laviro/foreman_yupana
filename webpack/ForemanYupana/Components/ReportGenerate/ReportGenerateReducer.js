import Immutable from 'seamless-immutable';

import {
  REPORT_GENERATE_PROCESS_START,
  REPORT_GENERATE_PROCESS_STOP,
  REPORT_GENERATE_PROCESS_FINISH,
} from './ReportGenerateConstants';

const initialState = Immutable({
  /** insert ReportGenerate state here */
  processID: 0,
  status: 'stopped',
});

export default (state = initialState, action) => {
  const { payload: { processID, status } = {} } = action;
  switch (action.type) {
    case REPORT_GENERATE_PROCESS_START:
      return state.merge({
        processID,
        status,
      });
    case REPORT_GENERATE_PROCESS_STOP:
      return state.merge({
        status,
      });
    case REPORT_GENERATE_PROCESS_FINISH:
      return state.merge({
        status,
      });
    default:
      return state;
  }
};
