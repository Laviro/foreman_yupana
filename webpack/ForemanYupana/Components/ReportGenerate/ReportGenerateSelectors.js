import { selectForemanYupana } from '../../ForemanYupanaSelectors';
import { selectGenerating } from '../Dashboard/DashboardSelectors';

export const selectReportGenerate = state =>
  selectForemanYupana(state).generating;
export const selectProcessID = state => selectReportGenerate(state).processID;
export const selectLogs = state => selectGenerating(state).logs;
export const selectCompleted = state => selectGenerating(state).completed;
export const selectError = state => selectGenerating(state).error;
export const selectLoading = state =>
  selectReportGenerate(state).status === 'running';

export const selectExitCode = state => {
  const { status } = selectReportGenerate(state);
  let exitCode;
  switch (status) {
    case 'success':
      exitCode = '0';
      break;
    case 'failure':
      exitCode = '1';
      break;
    default:
      exitCode = '--';
  }
  return exitCode;
};
