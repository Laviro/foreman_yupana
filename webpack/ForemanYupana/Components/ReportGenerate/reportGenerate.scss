.report-generate {
  padding: 20px;

  .row {
    margin-bottom: 20px;
  }

  .tab-header {
    button {
      float: right;
    }
  }

  .terminal {
    height: 300px;
    background-color: #3a3a3a;
    border: 5px solid #aaaaaa;
    padding: 10px 0;

    p {
      color: white;
    }
  }

  .tab-footer {
    p {
      span {
        margin-right: 5px;
      }
    }
  }
}
