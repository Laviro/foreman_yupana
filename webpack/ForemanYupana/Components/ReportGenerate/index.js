import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import * as actions from './ReportGenerateActions';
import reducer from './ReportGenerateReducer';

import ReportGenerate from './ReportGenerate';
import {
  selectLogs,
  selectProcessID,
  selectCompleted,
  selectLoading,
  selectExitCode,
} from './ReportGenerateSelectors';
import { selectError } from '../ReportUpload/ReportUploadSelectors';

// map state to props
const mapStateToProps = state => ({
  /** add state keys here */
  processID: selectProcessID(state),
  logs: selectLogs(state),
  completed: selectCompleted(state),
  loading: selectLoading(state),
  exitCode: selectExitCode(state),
  error: selectError(state),
});

// map action dispatchers to props
const mapDispatchToProps = dispatch => bindActionCreators(actions, dispatch);

// export reducers
export const reducers = { generating: reducer };

// export connected component
export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ReportGenerate);
