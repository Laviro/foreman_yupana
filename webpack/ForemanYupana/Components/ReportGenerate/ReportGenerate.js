import React from 'react';
import PropTypes from 'prop-types';
import { noop } from 'patternfly-react';
import TabContainer from '../TabContainer';
import TabHeader from '../TabHeader';
import TabFooter from '../TabFooter';
import Terminal from '../Terminal';
import ScheduledRun from '../ScheduledRun';
import './reportGenerate.scss';

const ReportGenerate = ({
  exitCode,
  logs,
  onRestart,
  processScheduledTime,
}) => (
  <TabContainer className="report-generate">
    <TabHeader exitCode={exitCode} onRestart={onRestart} />
    <Terminal>{logs}</Terminal>
    <TabFooter>
      <ScheduledRun time={processScheduledTime} />
    </TabFooter>
  </TabContainer>
);

ReportGenerate.propTypes = {
  exitCode: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  logs: PropTypes.arrayOf(PropTypes.string),
  onRestart: PropTypes.func,
  processScheduledTime: PropTypes.string,
};

ReportGenerate.defaultProps = {
  exitCode: 0,
  logs: ['No running process'],
  onRestart: noop,
  processScheduledTime: '00:00',
};

export default ReportGenerate;
