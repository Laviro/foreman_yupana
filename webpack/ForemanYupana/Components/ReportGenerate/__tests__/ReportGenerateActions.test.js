import { testActionSnapshotWithFixtures } from 'react-redux-test-utils';
import {
  startProcess,
  stopProcess,
  restartProcess,
} from '../ReportGenerateActions';
import { processID } from '../ReportGenerate.fixtures';
import { yupana } from '../../Dashboard/Dashboard.fixtures';

const fixtures = {
  'should startProcess': () => startProcess(),
  'should stopProcess': () => stopProcess(processID),
  'should restartProcess': () => restartProcess(),
};

beforeEach(() => {
  window.__yupana__ = yupana;
});

describe('ReportGenerate actions', () =>
  testActionSnapshotWithFixtures(fixtures));
