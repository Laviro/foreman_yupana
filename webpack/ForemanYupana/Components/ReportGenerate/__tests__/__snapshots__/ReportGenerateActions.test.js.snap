// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReportGenerate actions should restartProcess 1`] = `
Object {
  "type": "YUPANA/REPORT_GENERATE_PROCESS_RESTART",
}
`;

exports[`ReportGenerate actions should startProcess 1`] = `
Array [
  Array [
    Object {
      "payload": Object {
        "processID": 3,
        "status": "running",
      },
      "type": "YUPANA/REPORT_GENERATE_PROCESS_START",
    },
  ],
]
`;

exports[`ReportGenerate actions should stopProcess 1`] = `
Array [
  Array [
    Object {
      "payload": Object {
        "status": "stopped",
      },
      "type": "YUPANA/REPORT_GENERATE_PROCESS_STOP",
    },
  ],
]
`;
